SQL: Add method args to PERCENTILE/PERCENTILE_RANK

palesz opened the pull request:
* Adds the capability to have functions with two optional arguments
* Adds two new optional arguments to `PERCENTILE()` and
  `PERCENTILE_RANK()` functions, namely the method and
  method_parameter which can be: 1) `tdigest` and a double `compression`
  parameter or 2) `hdr` and an integer representing the
  `number_of_digits` parameter.
* Integration tests
* Documentation updates

Closes #63567

Commit:
SQL: Add method args to PERCENTILE/PERCENTILE_RANK

* Adds the capability to have functions with two optional arguments
* Adds two new optional arguments to `PERCENTILE()` and
  `PERCENTILE_RANK()` functions, namely the method and
  method_parameter which can be: 1) `tdigest` and a double `compression`
  parameter or 2) `hdr` and an integer representing the
  `number_of_digits` parameter.
* Integration tests
* Documentation updates

Closes #63567

elasticmachine commented:
Pinging @elastic/es-ql (:Query Languages/SQL)

Commit:
Reverted the FunctionRegistry refactor and comments

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java:
```diff
@@ -45,6 +46,29 @@ public static TypeResolution isIP(Expression e, String operationName, ParamOrdin
         return isType(e, dt -> dt == IP, operationName, paramOrd, "ip");
     }
 
+    public static TypeResolution isEnum(Expression e, String operationName, ParamOrdinal paramOrd, Set<String> acceptedValues) {
```
Unless this method is used in multiple places, I would not promote it to `TypeResolution`.
The name also looks incorrect, `isEnum` but there's no enum in the args - if I'm guessing the intent correctly, pass an `Enum` along side a translating function so the object/string mapping is generalized:

`<E extends Enum<E>> isStringInEnum(Expression e, Function<Enum<E>, String> value, Enum<E> e, String operationName, ParamOrdinal paramOrd)`

This however seems too complicated for a narrow case, just checking a string inside a list of strings is much more common but that should be reflected in the name.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java:
```diff
@@ -45,6 +46,29 @@ public static TypeResolution isIP(Expression e, String operationName, ParamOrdin
+            return resolution;
+        }
+
+        String value = (String)e.fold();
```
Yet again an issue with formatting - please fix it.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:
```diff
@@ -420,16 +420,23 @@ public static FunctionDefinition def(Class<? extends Function> function, Functio
-            if (hasMinimumThree && (children.size() > 4 || children.size() < 3)) {
-                throw new QlIllegalArgumentException("expects three or four arguments");
-            } else if (!hasMinimumThree && children.size() != 4) {
+            if (TwoOptionalArguments.class.isAssignableFrom(function)) {
+                if (children.size() > 4 || children.size() < 2) {
+                    throw new QlIllegalArgumentException("expects minimum two, maximum four arguments");
+                }
+            } else if (OptionalArgument.class.isAssignableFrom(function)) {
+                if (children.size() > 4 || children.size() < 3) {
+                    throw new QlIllegalArgumentException("expects three or four arguments");
+                }
+            } else if (children.size() != 4) {
```
The initial order is more readable - start with `OptionalArgument`, second `TwoOptionalArguments`

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java:
```diff
@@ -420,16 +420,23 @@ public static FunctionDefinition def(Class<? extends Function> function, Functio
             }
-            return ctorRef.build(source, children.get(0), children.get(1), children.get(2), children.size() == 4 ? children.get(3) : null);
+            return ctorRef.build(source, children.get(0), children.get(1),
+                children.size() > 2 ? children.get(2) : null,
```
Extract the children extraction the `if`s above - it's easier to understand what branch did what.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/Functions.java:
```diff
@@ -18,4 +18,12 @@ public static boolean isAggregate(Expression e) {
-}
\ No newline at end of file
+
+    public static int countOfNonNullOptionalArgs(Object... args) {
+        int shift = 0;
+        for (Object a : args) {
+            shift += a == null ? 0 : 1;
+        }
+        return shift;
+    }
```
This method is not tied to functions, has a clunky name and bad names - countOf -> return shift?.
Please remove it.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Percentile.java:
```diff
@@ -16,31 +17,44 @@
-    public Percentile(Source source, Expression field, Expression percent) {
-        super(source, field, singletonList(percent));
+    public Percentile(Source source, Expression field, Expression percent, Expression method, Expression methodParameter) {
+        super(source, field, Stream.of(percent, (method = defaultMethod(source, method)),
+            (methodParameter = defaultMethodParameter(methodParameter))).filter(Objects::nonNull).collect(Collectors.toList()));
```
Instead of an in-line stream creation/filtering/collection inside a super invocation (a no-no), extract this into a static utility method.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Percentile.java:
```diff
@@ -16,31 +17,44 @@
     public Percentile replaceChildren(List<Expression> newChildren) {
-        if (newChildren.size() != 2) {
-            throw new IllegalArgumentException("expected [2] children but received [" + newChildren.size() + "]");
+        if (children().size() != newChildren.size()) {
```
This should be static check not a dynamic one.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java:
```diff
@@ -0,0 +1,119 @@
+        String sourceText, Expression method, Expressions.ParamOrdinal methodOrdinal,
+        Expression methodParameter, Expressions.ParamOrdinal methodParameterOrdinal) {
+
+        Expression.TypeResolution resolution = isEnum(method, sourceText, methodOrdinal, ACCEPTED_METHODS);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (methodParameter != null) {
+            resolution = isFoldable(methodParameter, sourceText, methodParameterOrdinal);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+
+            PercentilesMethod percentilesMethod = asPercentilesMethod(method);
+            switch (percentilesMethod) {
+                case TDIGEST:
+                    resolution = isNumeric(methodParameter, sourceText, methodParameterOrdinal);
+                    break;
+                case HDR:
+                    resolution = isInteger(methodParameter, sourceText, methodParameterOrdinal);
+                    break;
+                default:
+                    throw new QlIllegalArgumentException("Not handled PercentilesMethod [" + percentilesMethod + "]");
+            }
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
```
Why is there a need for `isEnum` if the methodParam is checked again to be folded and then checked against the enum values itself?
Further more the point of type resolution is to *not* throw exceptions but return error messages.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java:
```diff
@@ -0,0 +1,119 @@
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isInteger;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isNumeric;
+
+public abstract class PercentileMethodConfiguration {
```
This class needs rework - it's hard to understand its purpose and style. The code both goes multiple checks and does either enum to string or directly on string validation.
There should be only one way and if there's a dedicated class, make that customized for the enum it tries to handle.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T foldOptionalNullable(Expression e, DataType dataType) {
+        if (e == null) {
+            return (T) null;
+        }
+        return (T) SqlDataTypeConverter.convert(Foldables.valueOf(e), dataType);
+    }
+
+    private static Expression nullIfUnspecified(Expression e) {
+        return e == null || (e.foldable() && e.fold() == null) ? null : e;
+    }
```
Two methods doing similar checks for null and folding.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1008,37 +1008,34 @@ public LogicalPlan apply(LogicalPlan p) {
-            });
+            // create a Percentile agg for each argument combination
+            Map<List<Expression>, Percentiles> percentilesPerArgs = new LinkedHashMap<>();
+            percentsPerArgs.forEach((arguments, percents) -> percentilesPerArgs.put(
```
arguments -> args

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1049,37 +1046,33 @@ public LogicalPlan apply(LogicalPlan p) {
-                ranksPerField.put(k, new PercentileRanks(v.iterator().next().source(), k, new ArrayList<>(v)));
-            });
+            Map<List<Expression>, PercentileRanks> ranksPerArgs = new LinkedHashMap<>();
+            valuesPerArgs.forEach((arguments, values) -> ranksPerArgs.put(
```
Readjust the comments.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1049,37 +1046,33 @@ public LogicalPlan apply(LogicalPlan p) {
-            });
+            Map<List<Expression>, PercentileRanks> ranksPerArgs = new LinkedHashMap<>();
+            valuesPerArgs.forEach((arguments, values) -> ranksPerArgs.put(
+                arguments,
```
arguments -> args

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1049,37 +1046,33 @@ public LogicalPlan apply(LogicalPlan p) {
-                    }
-
-                    percentiles.add(per.value());
+                    valuesPerArgs.computeIfAbsent(keyOf(per), k -> new LinkedHashSet<>()).add(per.value());
```
👍

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1008,37 +1008,34 @@ public LogicalPlan apply(LogicalPlan p) {
-                    }
-
-                    percentiles.add(per.percent());
+                    percentsPerArgs.computeIfAbsent(keyOf(per), k -> new LinkedHashSet<>()).add(per.percent());
```
👍

costin reviewed (changes requested):
The hard part, the optimizer and tests look good. The style around validation needs work - the code is convoluted and large making it hard to follow.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java:
```diff
@@ -0,0 +1,119 @@
+            }
+        }
+        if (percentilesMethod == null) {
+            throw new QlIllegalArgumentException("specified method [ " + methodName + " ] is not one of the allowed methods "
```
I think here you need to not throw an exception, but create a `TypeResolution` to express the inconsistency.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileMethodConfiguration.java:
```diff
@@ -0,0 +1,119 @@
+                Integer numOfDigits = foldOptionalNullable(methodParameter, DataTypes.INTEGER);
+                return numOfDigits == null ? new PercentilesConfig.Hdr() : new PercentilesConfig.Hdr(numOfDigits);
+            default:
+                throw new QlIllegalArgumentException("Not handled PercentilesMethod [" + percentilesMethod + "]");
```
Here the same. About `TypeResolution` instead of Exception.

astefan commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java:
```diff
@@ -45,6 +46,29 @@ public static TypeResolution isIP(Expression e, String operationName, ParamOrdin
+
+        String value = (String)e.fold();
+        if (acceptedValues.contains(value) == false) {
+            return new TypeResolution(format(null, "{}argument of [{}] must be a string, one of {}, received [{}]",
```
The error message is a bit messed up. This is how it looks with actual values in it: `third argument of [PERCENTILE(int, 50, null)] must be a string, one of [tdigest, hdr], received [null]`. How about `third argument of [PERCENTILE(int, 50, null)] must be one of [tdigest, hdr], received [null]`? To me it seems more clear and concise.

astefan reviewed (commented):
Left some comments.

Commit:
Refactor, PR suggestions

* Removed the `method` and `methodParameter` from the `Node.children()
`s (this way it also stops interacting with `FoldNull` unnecessarily)
* Added examples for the documentation
* Link the docs to the `percentiles()` aggregation documentation about
approximations
* Added extra testcases
* Folds everything that results with the same `PercentileConfig` into a
single aggregation.
* Reverted the whitespace changes

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

palesz reviewed (commented):
Thanks for the suggestions. Tried to take care all the suggestions.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Percentile.java:
```diff
@@ -72,4 +103,74 @@ public String innerName() {
+                } else {
+                    throw new IllegalStateException("Not handled PercentilesMethod [" + methodName + "], type resolution needs fix");
+                }
+                if (resolution.unresolved()) {
+                    return resolution;
+                }
```
seems superfluous.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/PercentileRanksAgg.java:
```diff
@@ -5,24 +5,30 @@
     private final List<Double> values;
+    private final PercentilesConfig percentilesConfig;
+
+    public static PercentilesConfig DEFAULT_PERCENTILES_CONFIG =
+        AggregationBuilders.percentileRanks("not_important", new double[1]).percentilesConfig();
```
superfluous?

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -2338,4 +2345,138 @@ public void testInInRangeValues() {
+        }).collect(Collectors.toList());
+    }
+
+    public void testPercentileNullOrUnspecifiedMethodOrDefaultMethodParameterIsTheSame() {
```
not sure if this had been the target of a previous comment, but I would myself deduplicate the tests that only differ in the name of the function (like this and `testPercentileRankNullOrUnspecifiedMethodOrDefaultMethodParameterIsTheSame` or `testPercentileOptimization` and `testPercentileRankOptimization`) with a simple two steps loop.

bpintea reviewed (approved):
nice work, lgtm, only left some minor optional comments.

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/docs/docs.csv-spec:
```diff
@@ -1511,6 +1511,26 @@ null           |6249.916666666667
+---------------+-----------------+-----------------
+null           |74999.0          |74992.0
+1              |72790.5          |70000.0
+2              |71924.70000000001|69936.0
+3              |73638.25         |71152.0
+4              |72115.59999999999|68464.0
```
Would different parameters results slightly better results (as in less digits)?

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/docs/docs.csv-spec:
```diff
@@ -1542,6 +1562,26 @@ null           |66.91240875912409
+
+   languages   |   rank_TDigest  |     rank_HDR
+---------------+-----------------+-----------------
+null           |73.65766569962062|80.0
+1              |73.7291625157734 |73.33333333333333
+2              |88.88005607010643|89.47368421052632
+3              |79.43662623295829|76.47058823529412
+4              |85.70446389643493|83.33333333333333
+5              |100.0            |95.23809523809524
```
Same here - wrap the results with a `ROUND` or potentially use different args to have only 3 decimals vs 14.

jrodewig commented on docs/reference/sql/functions/aggs.asciidoc:
```diff
@@ -477,14 +477,18 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[aggMadScalars]
 <1> a numeric field
-<2> a numeric expression (must be a constant and not based on a field)
+<2> a numeric literal
+<3> optional string literal to indicate the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
```
Current copy is fine as is, but this is a little more clear.
```suggestion
<3> optional string literal for the
<<search-aggregations-metrics-percentile-aggregation-approximation,percentile
algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
```

jrodewig commented on docs/reference/sql/functions/aggs.asciidoc:
```diff
@@ -477,14 +477,18 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[aggMadScalars]
-<2> a numeric expression (must be a constant and not based on a field)
+<2> a numeric literal
+<3> optional string literal to indicate the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
+<4> optional numeric literal for configuring the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Configures `compression` for `tdigest` or `number_of_significant_value_digits` for `hdr`. The default is the same as that of the backing algorithm.
```
Based on Andrei's review, `integer` may be a good replacement for `numeric literal`? I'd also streamline the wording slightly and remove the repeated link.
```suggestion
<4> optional integer that configures the percentile algorithm.
Configures `compression` for `tdigest` or `number_of_significant_value_digits`
for `hdr`. Uses the same defaults as the percentile algorithm.
```

jrodewig commented on docs/reference/sql/functions/aggs.asciidoc:
```diff
@@ -503,21 +507,31 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[aggPercentile]
 <1> a numeric field
-<2> a numeric expression (must be a constant and not based on a field)
+<2> a numeric literal
+<3> optional string literal to indicate the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
```
Same as above.
```suggestion
<3> optional string literal for the
<<search-aggregations-metrics-percentile-aggregation-approximation,percentile
algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
```

Commit:
Code cleanup, doc cleanup

jrodewig commented on docs/reference/sql/functions/aggs.asciidoc:
```diff
@@ -503,21 +507,31 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[aggPercentile]
-<2> a numeric expression (must be a constant and not based on a field)
+<2> a numeric literal
+<3> optional string literal to indicate the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
+<4> optional numeric literal for configuring the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Configures `compression` for `tdigest` or `number_of_significant_value_digits` for `hdr`. The default is the same as that of the backing algorithm.
```
Same as above.
```suggestion
<4> optional integer that configures the percentile algorithm.
Configures `compression` for `tdigest` or `number_of_significant_value_digits`
for `hdr`. Uses the same defaults as the percentile algorithm.
```

jrodewig reviewed (approved):
Aside from the points @astefan raised, the docs LGTM. I left some suggestions to streamline the wording, but feel free to ignore those if wanted. Thanks @palesz!

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentileConfig.java:
```diff
@@ -0,0 +1,62 @@
+
+    Expression methodParameter();
+
+    default PercentilesConfig percentileConfig() {
```
Typo, fixed.

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Percentile.java:
```diff
@@ -5,42 +5,58 @@
     @Override
     public Percentile replaceChildren(List<Expression> newChildren) {
-        if (newChildren.size() != 2) {
+        if (children().size() != 2) {
```
👍  good catch

palesz commented on docs/reference/sql/functions/aggs.asciidoc:
```diff
@@ -477,14 +477,18 @@ include-tagged::{sql-specs}/docs/docs.csv-spec[aggMadScalars]
-<2> a numeric expression (must be a constant and not based on a field)
+<2> a numeric literal
+<3> optional string literal to indicate the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Possible values: `tdigest` or `hdr`. Defaults to `tdigest`.
+<4> optional numeric literal for configuring the <<search-aggregations-metrics-percentile-aggregation-approximation,percentile algorithm>>. Configures `compression` for `tdigest` or `number_of_significant_value_digits` for `hdr`. The default is the same as that of the backing algorithm.
```
Same as above, technically `<3>` and `<4>` can be expressions too.

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileRank.java:
```diff
@@ -8,40 +8,45 @@
     public Expression replaceChildren(List<Expression> newChildren) {
-        if (newChildren.size() != 2) {
-            throw new IllegalArgumentException("expected [2] children but received [" + newChildren.size() + "]");
+        if (children().size() < 2) {
```
👍  yes, good catch

Commit:
More cleanups based on PR

astefan reviewed (commented):
Looking good. Just few more comments. Mostly cosmetic.

costin commented on x-pack/plugin/sql/qa/server/src/main/resources/docs/docs.csv-spec:
```diff
@@ -1511,6 +1511,26 @@ null           |6249.916666666667
+FROM emp 
+GROUP BY languages;
+
+   languages   | 97.3_TDigest  |   97.3_HDR    
+---------------+---------------+---------------
+null           |74999.0        |74992.0        
+1              |73717.0        |73712.0        
+2              |73530.238      |69936.0        
+3              |74970.0        |74992.0        
+4              |74572.0        |74608.0        
+5              |66117.118      |56368.0        
+// end::aggPercentileWithPercentileConfig
```
👍

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+import static org.elasticsearch.common.logging.LoggerMessageFormat.format;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isFoldable;
+
+public abstract class HasPercentilesConfig extends NumericAggregate implements EnclosedAgg, TwoOptionalArguments {
```
👍  Or just `PercentileAggregate`. The class should be protected, same as the constructor.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+            && Objects.equals(methodParameter, that.methodParameter);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), method, methodParameter);
+    }
```
hashcode before equals

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+    ));
+
+    // list of all the possible PercentileMethods that can we are capable of parameterizing as of now
+    private static final Map<PercentilesMethod, MethodParameterResolver> METHOD_TO_RESOLVER = Map.of(
```
`Map.of` will create issues when backporting to 7.11

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+        Expression.TypeResolution resolve(Expression methodParameter, String sourceText, Expressions.ParamOrdinal methodParameterOrdinal);
+    }
+
+    private static Expression.TypeResolution resolvePercentileConfiguration(
```
There's no value in making this method static - in fact it hurts. The method/params should be taken from the enclosing class simplifying the signature.
The `NAME_TO_METHOD` (has little meaning to say `ENUM_RESOLUTION`) can contain `Function<Enum,TypeResolution>` and the declaration can just pick up the arguments from the enclosing context inside the lambda.

palesz replied:
Inlined the static method. 

About `NAME_TO_METHOD`: Lambda can only pick up the fields `method` and `methodParameter` if the `NAME_TO_METHOD` is not static, otherwise the it has no access to the instance fields. Making it non-static on the other hand feels a bit wasteful.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
```
If super.equals is called, no need to call `==` or do the getClass() comparison since the super does it already.

palesz replied:
True. Remove. On the same note: `Function.equals()` seems copy of `Node.equals()`, I will remove that in a cleanup PR separately.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Percentile.java:
```diff
@@ -21,26 +20,26 @@
+        return new Percentile(source(), newChildren.get(0), newChildren.get(1), method(), methodParameter());
     }
 
     @Override
```
I believe this method is the same for Percentile/PercentileRanks so it can be extracted into the parent.

palesz replied:
See https://github.com/elastic/elasticsearch/pull/65026#discussion_r526953312

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileRanks.java:
```diff
@@ -5,35 +5,66 @@
+        return Objects.equals(percentilesConfig, that.percentilesConfig);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(super.hashCode(), percentilesConfig);
+    }
+
```
again, hashcode before equals

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileRanks.java:
```diff
@@ -5,35 +5,66 @@
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        if (!super.equals(o)) {
+            return false;
+        }
```
by calling super, the getClass() and == check are redundant.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1006,39 +1009,46 @@ public LogicalPlan apply(LogicalPlan p) {
         }
     }
 
+    private static class PercentileAggKey extends Tuple<Expression, PercentilesConfig> {
+        PercentileAggKey(HasPercentilesConfig per) {
+            super(per.field(), per.percentilesConfig());
+        }
+        
+        public Expression field() {
+            return v1();
+        }
+        
+        public PercentilesConfig percentilesConfig() {
+            return v2();
+        }
+    }
```
Percentile is already an agg making `PercentileAgg` redundant - just call the tuple `PercentileKey`.
Also the methods should be all private/protected where possible.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T foldOptionalNullable(Expression e, DataType dataType) {
+        if (e == null) {
+            return null;
+        }
+        return (T) SqlDataTypeConverter.convert(Foldables.valueOf(e), dataType);
+    }
```
return e == null ?  null : SqlDataTypeConverter.convert(Foldables.valueOf(e), dataType);

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> T foldOptionalNullable(Expression e, DataType dataType) {
```
The name is misleading since nullable means something else. rename to `foldNullSafe`

costin reviewed (changes requested):
Improved but there are still some loose ends, incorrect formatting and improper names.

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/HasPercentilesConfig.java:
```diff
@@ -0,0 +1,189 @@
+                MethodParameterResolver resolver = METHOD_TO_RESOLVER.get(percentilesMethod);
+                if (resolver == null) {
+                    // so in the future if a new method is added, at least the users will be able to use it with 
+                    // the default parameters, but won't be able to configure it until the resolver is added
```
Sadly, I have to remove it. Looking at it more we cannot future proof automagically. The `asPercentilesConfig` won't be able to translate the enum to a `PercentilesConfig` class anyways (we will hit the `default` case that will result in IllegalArgumentException). The best I can do is to add a unit test whether all the methods are handled, but server team might not be happy about that.

palesz commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileRank.java:
```diff
@@ -22,26 +19,26 @@
     public Expression replaceChildren(List<Expression> newChildren) {
-        if (newChildren.size() != 2) {
-            throw new IllegalArgumentException("expected [2] children but received [" + newChildren.size() + "]");
+        if (newChildren.size() < 2) {
```
I will keep this check in the `Percentile` and `PercentileRank` classes. The `HasPercentilesConfig` class does not have explicit static (rather only dynamic) knowledge about the children count of the child classes. Right now it just happens to be that only Percentile and PercentileRank are the child classes, each with two children.

Commit:
More cleanups based on PR

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

Commit:
More cleanups based on PR

Commit:
Using LinkedHashMap instead of normal map

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

astefan reviewed (approved):
Looking good. Great job! LGTM

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/PercentileRanks.java:
```diff
@@ -5,35 +5,64 @@
 import java.util.List;
+import java.util.Objects;
 
 public class PercentileRanks extends CompoundNumericAggregate {
```
Considering this file and Percentiles have grew in size and have two properties plus the associated equals/hashCode is worth extracting the functionlity into a common base class that extends CompoundNumericAggregate and just like their function equivalent, the actual implementations only inherit and declare the values/fields as a synonim to the internal expressions list and handle the info and replaceChildren methods.

palesz replied:
While implementing this suggested change, found the following:

```java
    public InnerAggregate(Source source, AggregateFunction inner, CompoundAggregate outer, Expression innerKey) {
        super(source, outer.field(), outer.arguments());
```

Where the `outer.arguments()` delegates to `children()` which already contains the `field()`, so the InnerAggregate will have two `field`s in the `children` list. Will address it separately, seemingly it does not cause any issues.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java:
```diff
@@ -1006,39 +1008,50 @@ public LogicalPlan apply(LogicalPlan p) {
     }
 
+    private static class PercentileKey extends Tuple<Expression, PercentilesConfig> {
+        PercentileKey(Percentile per) {
```
A small nit - to avoid having two constructors, it would be possible make the base class public so that percentilesConfig can be accessed. Since the rest of the methods are not, it keeps the encapsulation while removing redundancy.

costin reviewed (changes requested):
Looks good!

Commit:
Further PR comments

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support

Commit:
QL: Stop duplicating children in InnerAggregate

Instead of the `arguments()` that is the synonym of the `children()`,
use the `parameters()` (everything except the field) to avoid
duplicating the `field` in the `children` collection.

Commit:
Merge remote-tracking branch 'origin/master' into enhancement/percentile-hdr-support
